CMakeLists.txt

```diff
@@ -936,10 +936,16 @@ if (CMAKE_OSX_ARCHITECTURES STREQUAL "arm64" OR CMAKE_GENERATOR_PLATFORM_LWR STR
             list(APPEND ARCH_FLAGS -mfpu=neon-fp-armv8 -mno-unaligned-access)
         endif()
         if (${CMAKE_SYSTEM_PROCESSOR} MATCHES "armv7")
-            # Raspberry Pi 2
-            list(APPEND ARCH_FLAGS -mfpu=neon-fp-armv8 -mno-unaligned-access -funsafe-math-optimizations)
+            if ("${CMAKE_SYSTEM_NAME}" STREQUAL "Android")
+                # Android armeabi-v7a
+                list(APPEND ARCH_FLAGS -mfpu=neon-vfpv4 -mno-unaligned-access -funsafe-math-optimizations)
+            else()
+                # Raspberry Pi 2
+                list(APPEND ARCH_FLAGS -mfpu=neon-fp-armv8 -mno-unaligned-access -funsafe-math-optimizations)
+            endif()
```

Comment:
Is there a reason we are overriding -mfpu anyway? Shouldn't the default of -march=native cause the compiler to use the appropriate -mfpu by default?

Reply:
Should, I have no idea why the `auto` setting is overriden here as well.
I've crosscompiled for the Android armeabi-v7a with removed `-mfpu` flag and it compiled successfully as well. I've only set the flag to `neon-vfpv4` as suggested in the attached Issue (#4421). Maybe @prusnak could shed some light onto the meaning of those flags for arm v6 and v7?

Reply:
auto only works if you are compiling on the target system

specifying flags explicitly allows you to crosscompile for different system

Reply:
Based on [this](https://gcc.gnu.org/onlinedocs/gcc/ARM-Options.html#:~:text=The%20setting%20%E2%80%98auto%E2%80%99%20is%20the%20default%20and%20is%20special.%20It%20causes%20the%20compiler%20to%20select%20the%20floating%2Dpoint%20and%20Advanced%20SIMD%20instructions%20based%20on%20the%20settings%20of%20%2Dmcpu%20and%20%2Dmarch.) we should be able to replace `-mfpu` with `-march` and the compiler would do the rest for us. I do not own any of these Raspberries so I cannot even test this myself unfortunately.

Reply:
I don't think it should be the build script's job to guess what kind of FPU you have based only on CMAKE_SYSTEM_PROCESSOR and CMAKE_SYSTEM_NAME anyway - best to explicitly specify the necessary -march and -mcpu (and -mfpu if desired, but shouldn't be necessary) in your toolchain file's CMAKE_C_FLAGS_INIT/CMAKE_CXX_FLAGS_INIT.

Reply:
Question is then if we want to completely get rid of setting those flags?
I'm not that familiar with llama.cpp contribution guidelines so I do not know if there's any plan for such thing.
At the moment, I'd say it'd make sense given current build script structure.

Reply:
I suppose the problem is that Android builds are often cross-compiled (i.e. you almost never build on the actual device).

Agree it's a bit of a mess with these flags as it is (same in `whisper.cpp`). But since we lack all the various hardware to test so many cases, we can patch stuff one-by-one